tt_metal/impl/trace/trace.cpp

```diff
@@ -62,35 +62,31 @@ uint32_t Trace::next_id() {
     return global_trace_id++;
 }
 
-std::shared_ptr<TraceBuffer> Trace::create_trace_buffer(
-    const CommandQueue& cq, shared_ptr<detail::TraceDescriptor> desc, uint32_t unpadded_size) {
-    size_t page_size = interleaved_page_size(
-        unpadded_size, cq.device()->num_banks(BufferType::DRAM), kExecBufPageMin, kExecBufPageMax);
-    uint64_t padded_size = round_up(unpadded_size, page_size);
-
-    // Commit the trace buffer to device DRAM
+std::shared_ptr<TraceBuffer> Trace::create_empty_trace_buffer() {
     return std::make_shared<TraceBuffer>(
-        desc,
-        std::make_shared<Buffer>(
-            cq.device(), padded_size, page_size, BufferType::DRAM, TensorMemoryLayout::INTERLEAVED));
+        std::make_shared<detail::TraceDescriptor>(),
+        nullptr
+    );
 }
 
 void Trace::initialize_buffer(CommandQueue& cq, std::shared_ptr<TraceBuffer> trace_buffer) {
-    vector<uint32_t>& data = trace_buffer->desc->data;
-
-    uint64_t unpadded_size = data.size() * sizeof(uint32_t);
-    TT_FATAL(
-        unpadded_size <= trace_buffer->buffer->size(),
-        "Trace data size {} is larger than specified trace buffer size {}. Increase specified buffer size.",
-        unpadded_size,
-        trace_buffer->buffer->size());
-    size_t numel_padding = (trace_buffer->buffer->size() - unpadded_size) / sizeof(uint32_t);
+    vector<uint32_t>& trace_data = trace_buffer->desc->data;
+    uint64_t unpadded_size = trace_data.size() * sizeof(uint32_t);
+    size_t page_size = interleaved_page_size(
+        unpadded_size, cq.device()->num_banks(BufferType::DRAM), kExecBufPageMin, kExecBufPageMax);
+    uint64_t padded_size = round_up(unpadded_size, page_size);
+    size_t numel_padding = (padded_size - unpadded_size) / sizeof(uint32_t);
     if (numel_padding > 0) {
-        data.resize(data.size() + numel_padding, 0 /*padding value*/);
+        trace_data.resize(trace_data.size() + numel_padding, 0 /*padding value*/);
     }
-    uint64_t padded_size = data.size() * sizeof(uint32_t);
-    EnqueueWriteBuffer(cq, trace_buffer->buffer, data, kBlocking);
-
+    cq.device()->trace_buffers_size += padded_size;
+    TT_FATAL(
+        cq.device()->trace_buffers_size <= cq.device()->allocator_->config.trace_region_size,
+        "Creating trace buffers of size {}B on device {}, but only {}B is allocated for trace region.",  cq.device()->trace_buffers_size, cq.device()->id(),  cq.device()->allocator_->config.trace_region_size);
```

Comment:
Is this something we should be tracking/asserting for ourselves? Doesn't allocator already assert out if we go over the size? This also doesn't account for unfilled banks/fragmentation and would just fall through to the allocator assertions anyways so do we need another explicit assert here?

Reply:
The allocator asserts themselves aren't really descriptive. The goal here was to give users something similar to the assert we initially had in place, that informed the user when the trace buffer could not fit the actual trace.

Reply:
I see. I guess my concern with this was that since this doesn't cover all cases, users would get a mix of assertion errors for similar issues for trace allocation and that we are doing extra/duplicated bookkeeping to the allocator (fairly minimal though). Having a more descriptive assertion is useful though.